File: python/eou/chat_template.py
from __future__ import annotations

from dataclasses import dataclass

from . import constants

@dataclass
class Turn:
    role: str
    content: str

    @classmethod
    def user(cls, content: str) -> "Turn":
        return cls(role="user", content=content)

    @classmethod
    def assistant(cls, content: str) -> "Turn":
        return cls(role="assistant", content=content)

def format_qwen_chat(turns: list[Turn], partial: str) -> str:
    parts: list[str] = []
    for t in turns:
        role = t.role if t.role else "user"
        parts.append(constants.IM_START)
        parts.append(role)
        parts.append("\n")
        parts.append(t.content)
        parts.append(constants.IM_END)
        parts.append("\n")
    if partial:
        parts.append(constants.IM_START)
        parts.append("user\n")
        parts.append(partial)
    return "".join(parts)

def rolling_history(turns: list[Turn], max_turns: int) -> list[Turn]:
    if max_turns == 0 or len(turns) <= max_turns:
        return turns
    return turns[len(turns) - max_turns :]

File: rust/src/eou/chat_template.rs
#![allow(dead_code)]

use crate::defaults;

#[derive(Clone, Debug)]
pub struct Turn {
    pub role: String,
    pub content: String,
}

impl Turn {
    pub fn user(content: impl Into<String>) -> Self {
        Self {
            role: "user".into(),
            content: content.into(),
        }
    }

    pub fn assistant(content: impl Into<String>) -> Self {
        Self {
            role: "assistant".into(),
            content: content.into(),
        }
    }
}

pub fn format_qwen_chat(turns: &[Turn], partial: &str) -> String {
    let mut s = String::new();
    for t in turns {
        let role = if t.role.is_empty() {
            "user"
        } else {
            t.role.as_str()
        };
        s.push_str(defaults::eou::IM_START);
        s.push_str(role);
        s.push('\n');
        s.push_str(&t.content);
        s.push_str(defaults::eou::IM_END);
        s.push('\n');
    }
    if !partial.is_empty() {
        s.push_str(defaults::eou::IM_START);
        s.push_str("user\n");
        s.push_str(partial);
    }
    s
}

pub fn rolling_history(turns: &[Turn], max_turns: usize) -> &[Turn] {
    if max_turns == 0 || turns.len() <= max_turns {
        return turns;
    }
    &turns[turns.len() - max_turns..]
}
